In [1]:
import pandas as pd
df = pd.read_csv('../../dataset/final/merged_dataset_final_all_pp.csv')

In [2]:
import pandas as pd

# Calculate dataset-based thresholds
tp7_threshold = df['tp_7d_cum'].quantile(0.75)      # top 25% cumulative rainfall
tp_lag3_threshold = df['tp_lag3'].quantile(0.25)    # lowest 25% dry period
tp_mm_threshold = df['tp_mm'].quantile(0.75)        # sudden rainfall top 25%
heat_threshold = 27                                   # 9 AM heatwave threshold in KTM
low_n_threshold = df['total_nitrogen'].quantile(0.25)  # low N
low_ph_threshold = df['ph'].quantile(0.25)             # acidic soil

In [3]:
def rice_blast_risk_pct(row):
    risk = 0
    # Temperature favorable
    if 24 <= row['t2m_C'] <= 30:
        risk += 1
    if row['t2m_C'] > heat_threshold:
        risk += 1
    # Rainfall / moisture
    if row['tp_7d_cum'] > tp7_threshold:
        risk += 1
    if row['consec_rain_days'] >= 2:
        risk += 1
    if row['tp_lag3'] < tp_lag3_threshold and row['tp_mm'] > tp_mm_threshold:
        risk += 2  # sudden rain after dry period
    # Soil
    if row['total_nitrogen'] < low_n_threshold:
        risk += 1
    if row['ph'] < low_ph_threshold:
        risk += 1

    # Risk classification
    if risk >= 3:
        return "High"
    elif risk >= 2:
        return "Moderate"
    else:
        return "Low"

In [4]:
# 2️⃣ Bacterial Leaf Blight - percentile-based
def blb_risk_pct(row):
    risk = 0
    if 24 <= row['t2m_C'] <= 34:
        risk += 1
    if row['t2m_C'] > heat_threshold:
        risk += 1
    if row['tp_7d_cum'] > tp7_threshold:
        risk += 1
    if row['consec_rain_days'] >= 2:
        risk += 1
    if row['tp_lag3'] < tp_lag3_threshold and row['tp_mm'] > tp_mm_threshold:
        risk += 2
    if row['total_nitrogen'] < low_n_threshold:
        risk += 1
    if row['ph'] < low_ph_threshold:
        risk += 1

    if risk >= 3:
        return "High"
    elif risk >= 2:
        return "Moderate"
    else:
        return "Low"

In [5]:
# 3️⃣ Sheath Blight - percentile-based
def sheath_blight_risk_pct(row):
    risk = 0
    if row['t2m_C'] >= 24:
        risk += 1
    if row['t2m_C'] > heat_threshold:
        risk += 1
    if row['consec_rain_days'] >= 2:
        risk += 1
    if row['tp_7d_cum'] > tp7_threshold:
        risk += 1
    if row['tp_lag3'] > tp_lag3_threshold:
        risk += 2  # standing water proxy
    if row['total_nitrogen'] < low_n_threshold:
        risk += 1

    if risk >= 3:
        return "High"
    elif risk >= 2:
        return "Moderate"
    else:
        return "Low"

In [6]:
# Apply to dataset
df['blast_risk'] = df.apply(rice_blast_risk_pct, axis=1)
df['blb_risk'] = df.apply(blb_risk_pct, axis=1)
df['sheath_risk'] = df.apply(sheath_blight_risk_pct, axis=1)

# Composite rice disease risk
def rice_disease_risk_pct(row):
    if "High" in [row['blast_risk'], row['blb_risk'], row['sheath_risk']]:
        return "High"
    elif "Moderate" in [row['blast_risk'], row['blb_risk'], row['sheath_risk']]:
        return "Moderate"
    else:
        return "Low"

df['rice_disease_risk'] = df.apply(rice_disease_risk_pct, axis=1)

In [7]:
df.columns

Index(['latitude', 'longitude', 'nearest_lat', 'nearest_lon', 'distance',
       'year', 'month', 'day', 'tp_mm', 't2m_C', 'anomaly_T2m_C',
       'heat_stress_proxy', 'tp_7d_cum', 'tp_14d_cum', 'tp_7d_avg',
       'consec_rain_days', 'tp_lag1', 'tp_lag3', 'tp_lag7', 'heavy_rain',
       'month_sin', 'month_cos', 'heat_proxy', 'heat_next_day',
       'heat_next_2days', 'heat_next_3days', 'tp_anomaly', 'tp_std_anomaly',
       'heatwave_flag', 'next_day_match', 'next_2days_match',
       'next_3days_match', 'lat', 'lon', 'ph', 'organic_matter',
       'total_nitrogen', 'potassium', 'p2o5', 'boron', 'zinc', 'sand', 'clay',
       'slit', 'parentsoil', 'province', 'district', 'palika', 'crop',
       'variety', 'UREA1', 'UREA2', 'UREA3', 'DAP', 'MOP', 'organic',
       'boron_fert', 'palika_num', 'blast_risk', 'blb_risk', 'sheath_risk',
       'rice_disease_risk'],
      dtype='object')

In [8]:
df.head()

latitude  longitude  nearest_lat  nearest_lon  distance  year  month  day  \
0      27.5       85.0       27.652       85.005  0.152082  2021      1    1   
1      27.5       85.0       27.652       85.005  0.152082  2021      1    2   
2      27.5       85.0       27.652       85.005  0.152082  2021      1    3   
3      27.5       85.0       27.652       85.005  0.152082  2021      1    4   
4      27.5       85.0       27.652       85.005  0.152082  2021      1    5   

   tp_mm  t2m_C  ...  UREA3   DAP   MOP  organic  boron_fert  palika_num  \
0    0.0  19.51  ...   1.93  0.54  1.67      6.0         1.0         1.0   
1    0.0  20.18  ...   1.93  0.54  1.67      6.0         1.0         1.0   
2    0.0  20.59  ...   1.93  0.54  1.67      6.0         1.0         1.0   
3    0.0  18.81  ...   1.93  0.54  1.67      6.0         1.0         1.0   
4    0.0  20.24  ...   1.93  0.54  1.67      6.0         1.0         1.0   

   blast_risk  blb_risk  sheath_risk  rice_disease_risk  
0         Low       Low          Low                Low  
1         Low       Low          Low                Low  
2         Low       Low          Low                Low  
3         Low       Low          Low                Low  
4         Low       Low          Low                Low  

[5 rows x 62 columns]

In [23]:
# For each disease risk column
disease = ['blast_risk', 'blb_risk', 'sheath_risk', 'rice_disease_risk']

for col in disease:
    print(f"\nCounts for {col}:")
    print(df[col].value_counts())



Counts for blast_risk:
blast_risk
Low         18603
Moderate     5394
High         2471
Name: count, dtype: int64

Counts for blb_risk:
blb_risk
Low         18274
Moderate     5626
High         2568
Name: count, dtype: int64

Counts for sheath_risk:
sheath_risk
Low         10916
High         8299
Moderate     7253
Name: count, dtype: int64

Counts for rice_disease_risk:
rice_disease_risk
Low         10077
High         8703
Moderate     7688
Name: count, dtype: int64


In [24]:
df['tp_7d_cum'].max()

30.55906266

In [25]:
import folium
from IPython.display import display

# Palika-level mean coordinates
palika_coords = df.groupby('palika')[['nearest_lat','nearest_lon']].mean().reset_index()

# Risk counts per Palika
palika_risk = df.groupby('palika')['rice_disease_risk'].value_counts().unstack(fill_value=0).reset_index()

# Merge coordinates and risk
palika_map_df = palika_coords.merge(palika_risk, on='palika', how='left')

# Function to pick color based on risk
def risk_color(risk):
    if risk == 'High':
        return 'red'
    elif risk == 'Moderate':
        return 'orange'
    else:
        return 'green'

# Create Folium map
m = folium.Map(location=[27.7, 85.3], zoom_start=10)

# Add Palika markers
for _, row in palika_map_df.iterrows():
    if row['High'] > 0:
        risk_level = 'High'
    elif row['Moderate'] > 0:
        risk_level = 'Moderate'
    else:
        risk_level = 'Low'
        
    folium.CircleMarker(
        location=[row['nearest_lat'], row['nearest_lon']],
        radius=10,
        color=risk_color(risk_level),
        fill=True,
        fill_color=risk_color(risk_level),
        fill_opacity=0.7,
        popup=f"{row['palika']}\nHigh: {row['High']}, Moderate: {row['Moderate']}, Low: {row['Low']}\nRisk: {risk_level}"
    ).add_to(m)

# Display map directly in notebook
display(m)


KeyError: 'palika'

In [ ]:
df

latitude  longitude  nearest_lat  nearest_lon  distance  year  month  \
0          27.5       85.0       27.652       85.005  0.152082  2021      1   
1          27.5       85.0       27.652       85.005  0.152082  2021      1   
2          27.5       85.0       27.652       85.005  0.152082  2021      1   
3          27.5       85.0       27.652       85.005  0.152082  2021      1   
4          27.5       85.0       27.652       85.005  0.152082  2021      1   
...         ...        ...          ...          ...       ...   ...    ...   
26463      28.0       86.0       27.892       85.476  0.535014  2025     10   
26464      28.0       86.0       27.892       85.476  0.535014  2025     10   
26465      28.0       86.0       27.892       85.476  0.535014  2025     10   
26466      28.0       86.0       27.892       85.476  0.535014  2025     10   
26467      28.0       86.0       27.892       85.476  0.535014  2025     10   

       day  tp_mm  t2m_C  ...  UREA3   DAP   MOP  organic  boron_fert  \
0        1    0.0  19.51  ...   1.93  0.54  1.67      6.0         1.0   
1        2    0.0  20.18  ...   1.93  0.54  1.67      6.0         1.0   
2        3    0.0  20.59  ...   1.93  0.54  1.67      6.0         1.0   
3        4    0.0  18.81  ...   1.93  0.54  1.67      6.0         1.0   
4        5    0.0  20.24  ...   1.93  0.54  1.67      6.0         1.0   
...    ...    ...    ...  ...    ...   ...   ...      ...         ...   
26463   20    0.0  14.62  ...   1.81  2.17  1.67      6.0         1.0   
26464   21    0.0  14.62  ...   1.81  2.17  1.67      6.0         1.0   
26465   22    0.0  14.62  ...   1.81  2.17  1.67      6.0         1.0   
26466   23    0.0  14.62  ...   1.81  2.17  1.67      6.0         1.0   
26467   24    0.0  14.62  ...   1.81  2.17  1.67      6.0         1.0   

       palika_num  blast_risk  blb_risk  sheath_risk  rice_disease_risk  
0             1.0         Low       Low          Low                Low  
1             1.0         Low       Low          Low                Low  
2             1.0         Low       Low          Low                Low  
3             1.0         Low       Low          Low                Low  
4             1.0         Low       Low          Low                Low  
...           ...         ...       ...          ...                ...  
26463         NaN         Low       Low          Low                Low  
26464         NaN         Low       Low          Low                Low  
26465         NaN         Low       Low          Low                Low  
26466         NaN         Low       Low          Low                Low  
26467         NaN         Low       Low          Low                Low  

[26468 rows x 62 columns]

In [ ]:
for col in df.columns:
    print(f"{col}: {df[col].max()}")


latitude: 28.0
longitude: 86.0
nearest_lat: 27.894
nearest_lon: 85.957
distance: 0.9064325678173808
year: 2025
month: 12
day: 31
tp_mm: 15.577793
t2m_C: 35.37
anomaly_T2m_C: 16.552525812945714
heat_stress_proxy: 1
tp_7d_cum: 30.55906266
tp_14d_cum: 44.5885656
tp_7d_avg: 4.36558038
consec_rain_days: 7
tp_lag1: 15.577793
tp_lag3: 15.577793
tp_lag7: 15.577793
heavy_rain: 1
month_sin: 1.0
month_cos: 1.0
heat_proxy: 1
heat_next_day: 1
heat_next_2days: 1
heat_next_3days: 1
tp_anomaly: 14.652771112040922
tp_std_anomaly: 19.85568564291748
heatwave_flag: 1
next_day_match: True
next_2days_match: True
next_3days_match: True
lat: 27.894
lon: 85.957
ph: 6.23
organic_matter: 5.24
total_nitrogen: 0.24
potassium: 395.73
p2o5: 246.53
boron: 1.3
zinc: 2.75
sand: 57.16
clay: 29.74
slit: 56.61
parentsoil: 3.0
province: Bagmati
district: Sindhupalchok
palika: Tripurasundari Gaunpalika
crop: 0.0
variety: 1.0
UREA1: 2.21
UREA2: 3.62
UREA3: 3.62
DAP: 3.62
MOP: 3.34
organic: 6.0
boron_fert: 1.0
palika_num: 15.

In [26]:
df

latitude  longitude  nearest_lat  nearest_lon  distance  year  month  \
0          27.5       85.0       27.652       85.005  0.152082  2021      1   
1          27.5       85.0       27.652       85.005  0.152082  2021      1   
2          27.5       85.0       27.652       85.005  0.152082  2021      1   
3          27.5       85.0       27.652       85.005  0.152082  2021      1   
4          27.5       85.0       27.652       85.005  0.152082  2021      1   
...         ...        ...          ...          ...       ...   ...    ...   
26463      28.0       86.0       27.892       85.476  0.535014  2025     10   
26464      28.0       86.0       27.892       85.476  0.535014  2025     10   
26465      28.0       86.0       27.892       85.476  0.535014  2025     10   
26466      28.0       86.0       27.892       85.476  0.535014  2025     10   
26467      28.0       86.0       27.892       85.476  0.535014  2025     10   

       day  tp_mm  t2m_C  ...  UREA2  UREA3   DAP   MOP  organic  boron_fert  \
0        1    0.0  19.51  ...   1.93   1.93  0.54  1.67      6.0         1.0   
1        2    0.0  20.18  ...   1.93   1.93  0.54  1.67      6.0         1.0   
2        3    0.0  20.59  ...   1.93   1.93  0.54  1.67      6.0         1.0   
3        4    0.0  18.81  ...   1.93   1.93  0.54  1.67      6.0         1.0   
4        5    0.0  20.24  ...   1.93   1.93  0.54  1.67      6.0         1.0   
...    ...    ...    ...  ...    ...    ...   ...   ...      ...         ...   
26463   20    0.0  14.62  ...   1.81   1.81  2.17  1.67      6.0         1.0   
26464   21    0.0  14.62  ...   1.81   1.81  2.17  1.67      6.0         1.0   
26465   22    0.0  14.62  ...   1.81   1.81  2.17  1.67      6.0         1.0   
26466   23    0.0  14.62  ...   1.81   1.81  2.17  1.67      6.0         1.0   
26467   24    0.0  14.62  ...   1.81   1.81  2.17  1.67      6.0         1.0   

       blast_risk  blb_risk  sheath_risk  rice_disease_risk  
0             Low       Low          Low                Low  
1             Low       Low          Low                Low  
2             Low       Low          Low                Low  
3             Low       Low          Low                Low  
4             Low       Low          Low                Low  
...           ...       ...          ...                ...  
26463         Low       Low          Low                Low  
26464         Low       Low          Low                Low  
26465         Low       Low          Low                Low  
26466         Low       Low          Low                Low  
26467         Low       Low          Low                Low  

[26468 rows x 58 columns]

In [27]:
df.drop(columns='palika_num', inplace=True)
df.drop(columns='district', inplace=True)
df.drop(columns='province', inplace=True)
df.drop(columns='palika', inplace=True)




KeyError: "['palika_num'] not found in axis"

In [ ]:
null_cols = df.columns[df.isnull().any()]
df[null_cols].isnull().sum()


Series([], dtype: float64)

In [ ]:
df.dtypes

latitude             float64
longitude            float64
nearest_lat          float64
nearest_lon          float64
distance             float64
year                   int64
month                  int64
day                    int64
tp_mm                float64
t2m_C                float64
anomaly_T2m_C        float64
heat_stress_proxy      int64
tp_7d_cum            float64
tp_14d_cum           float64
tp_7d_avg            float64
consec_rain_days       int64
tp_lag1              float64
tp_lag3              float64
tp_lag7              float64
heavy_rain             int64
month_sin            float64
month_cos            float64
heat_proxy             int64
heat_next_day          int64
heat_next_2days        int64
heat_next_3days        int64
tp_anomaly           float64
tp_std_anomaly       float64
heatwave_flag          int64
next_day_match          bool
next_2days_match        bool
next_3days_match        bool
lat                  float64
lon                  float64
ph            

In [28]:
# from sklearn.preprocessing import StandardScaler

# # Separate features and scale numeric columns
# X = df.drop(columns=risk_cols)
# y_dict = {col: df[col] for col in risk_cols}  # store y separately

# numeric_cols = X.select_dtypes(include=['float64', 'int64']).columns
# scaler = StandardScaler()
# X[numeric_cols] = scaler.fit_transform(X[numeric_cols])

In [29]:
# from sklearn.model_selection import train_test_split

# # Use same split for all three diseases
# X_train, X_test = train_test_split(X, test_size=0.2, random_state=42, shuffle=True)

# y_train_dict = {}
# y_test_dict = {}
# for col in risk_cols:
#     y_train_dict[col], y_test_dict[col] = train_test_split(
#         y_dict[col], test_size=0.2, random_state=42, stratify=y_dict[col]
#     )


In [30]:
# from sklearn.ensemble import RandomForestClassifier

# rf_models = {}
# for col in risk_cols:
#     print(f"\n=== Training for {col} ===")
#     rf = RandomForestClassifier(
#         n_estimators=500,  # more trees for better training
#         max_depth=None,    # let it grow fully
#         random_state=42,
#         class_weight='balanced'
#     )
#     rf.fit(X_train, y_train_dict[col])
#     rf_models[col] = rf


In [31]:
# from sklearn.metrics import classification_report

# for col in risk_cols:
#     y_pred = rf_models[col].predict(X_test)
#     print(f"\n=== Classification Report for {col} ===")
#     print(classification_report(y_test_dict[col], y_pred))


*********************Algorithm Comparision

In [34]:
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from xgboost import XGBClassifier
from lightgbm import LGBMClassifier
from sklearn.model_selection import StratifiedKFold, cross_val_score
from sklearn.preprocessing import StandardScaler
from imblearn.over_sampling import SMOTE
from imblearn.pipeline import Pipeline
import numpy as np
import pandas as pd


In [37]:
# -------- Safe features --------
safe_features = [
    'latitude', 'longitude', 'nearest_lat', 'nearest_lon', 'distance',
    'year', 'month', 'day',
    'tp_mm', 't2m_C',
    'ph', 'organic_matter', 'total_nitrogen', 'potassium', 'p2o5',
    'boron', 'zinc', 'sand', 'clay', 'slit', 'parentsoil',
    'crop', 'variety',
    'UREA1', 'UREA2', 'UREA3', 'DAP', 'MOP', 'organic', 'boron_fert'
]

X = df[safe_features]


In [35]:
models_to_compare = {
    "Random Forest": RandomForestClassifier(
        n_estimators=300, max_depth=15, min_samples_split=10,
        min_samples_leaf=5, class_weight='balanced', random_state=42),
    
    "XGBoost": XGBClassifier(
        n_estimators=300, max_depth=10, learning_rate=0.1,
        subsample=0.8, colsample_bytree=0.8, random_state=42, n_jobs=-1),
    
    "LightGBM": LGBMClassifier(
        n_estimators=300, max_depth=10, learning_rate=0.1,
        class_weight='balanced', random_state=42),
    
    "Gradient Boosting": GradientBoostingClassifier(
        n_estimators=300, max_depth=5, random_state=42)
}


In [41]:
from sklearn.preprocessing import LabelEncoder  
skf = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)
results = []

numeric_cols = df[safe_features].select_dtypes(include=['float64', 'int64']).columns

for disease in ['blast_risk', 'blb_risk', 'sheath_risk']:
    print(f"\n=== Evaluating for {disease} ===")
    
    X = df[safe_features]
    y = df[disease].copy()
    
    # Encode string labels -> numeric
    le = LabelEncoder()
    y_encoded = le.fit_transform(y)
    
    for model_name, model in models_to_compare.items():
        pipeline = Pipeline(steps=[
            ('scaler', StandardScaler()),
            ('smote', SMOTE(random_state=42)),
            ('model', model)
        ])
        
        acc_scores = cross_val_score(pipeline, X, y_encoded, cv=skf, scoring='accuracy', n_jobs=-1)
        f1_scores = cross_val_score(pipeline, X, y_encoded, cv=skf, scoring='f1_weighted', n_jobs=-1)
        
        results.append({
            'Disease': disease,
            'Model': model_name,
            'Mean Accuracy': np.mean(acc_scores),
            'Mean F1': np.mean(f1_scores)
        })
        
        print(f"{model_name:20s} | Accuracy={np.mean(acc_scores):.4f} | F1={np.mean(f1_scores):.4f}")



=== Evaluating for blast_risk ===
Random Forest        | Accuracy=0.8598 | F1=0.8661
XGBoost              | Accuracy=0.9147 | F1=0.9147
LightGBM             | Accuracy=0.9148 | F1=0.9150


KeyboardInterrupt: 

### 🧪 Model Performance Comparison

| Disease       | Model              | Accuracy | F1 Score |
|----------------|--------------------|-----------|-----------|
| **blast_risk** | Random Forest       | 0.8589 | 0.8653 |
|                | XGBoost             | 0.9154 | 0.9155 |
|                | LightGBM            | 0.9146 | 0.9148 |
|                | Gradient Boosting   | 0.9017 | 0.9026 |
| **blb_risk**   | Random Forest       | 0.8666 | 0.8719 |
|                | XGBoost             | 0.9198 | 0.9198 |
|                | LightGBM            | 0.9194 | 0.9195 |
|                | Gradient Boosting   | 0.9050 | 0.9057 |
| **sheath_risk**| Random Forest       | 0.7392 | 0.7389 |
|                | XGBoost             | 0.8780 | 0.8781 |
|                | LightGBM            | 0.8815 | 0.8818 |
|                | Gradient Boosting   | 0.8685 | 0.8690 |


In [ ]:
joblib.dump(LGBMClassifier, f"../../models/disease_risk/{disease}_lightgbm_model.pkl")
joblib.dump(XGBClassifier, f"../../models/disease_risk/{disease}_XGBClassifier.pkl")
joblib.dump(GradientBoostingClassifier, f"../../models/disease_risk/{disease}_GradientBoostingClassifier.pkl")

joblib.dump(scaler, f"../../models/disease_risk/{disease}_scaler.pkl")

['../../models/disease_risk/sheath_risk_scaler.pkl']

: 